# 1. Import the meta data of items

In [183]:
import pickle
f = open('pickle_data/dict_all_items_info.p', 'rb')
dict_items_info = pickle.load(f)
f.close()

In [184]:
len(dict_items_info)

8464

In [3]:
dict_items_info.items()[0]

('02779-Q',
 {'color': 'Black/patterned',
  'image': {'item': 'http://lp.hm.com/hmprod?set=key[source],value[/model/2017/E00 0304576 020 14 3519.jpg]&set=key[rotate],value[]&set=key[width],value[]&set=key[height],value[]&set=key[x],value[]&set=key[y],value[]&set=key[type],value[STILL_LIFE_FRONT]&set=key[hmver],value[2]&call=url[file:/product/large]',
   'model': 'http://lp.hm.com/hmprod?set=key[source],value[/environment/2017/8LZ_0375_006R.jpg]&set=key[rotate],value[0]&set=key[width],value[3771]&set=key[height],value[4409]&set=key[x],value[0]&set=key[y],value[-11]&set=key[type],value[FASHION_FRONT]&set=key[hmver],value[0]&call=url[file:/product/large]'},
  'metricCategoryID': 'MEN_HM_MAN_TSHIRTS_VESTS_M226',
  'name': 'T-shirt',
  'url': 'http://www.hm.com/us/product/02779?article=02779-Q',
  'who': 'Men'})

# 2. Digging all kinds of categories

In [4]:
for serial in dict_items_info.keys()[0:10]:
    metricCategoryID = dict_items_info[serial]['metricCategoryID']
    # Delete the first and last strings
    categories = metricCategoryID.split('_')[1:-1] # without the first and the last
    print categories

['HM', 'MAN', 'TSHIRTS', 'VESTS']
['EVERYDAY', 'FASHION', 'DRESSES', 'JUMPSUITS']
['CASUAL', 'HOODIES', 'SWEATSHIRTS']
['DRESSES', 'JUMPSUITS', 'SHORT']
['HM', 'MAN', 'TSHIRTS', 'VESTS']
['TOPS', 'SHORT', 'SLEEVED']
['HM', 'MAN', 'TSHIRTS', 'VESTS']
['SHIRTS', 'BLOUSES', 'BLOUSES']
['HM', 'MAN', 'TSHIRTS', 'VESTS']
['HM', 'MAN', 'TSHIRTS', 'VESTS']


In [5]:
set_category = set()
set_category_first = set()
set_category_last = set()
for serial in dict_items_info:
    metricCategoryID = dict_items_info[serial]['metricCategoryID']
    # Delete the first and last strings
    categories = metricCategoryID.split('_')[1:-1]
    set_category_first.add(categories[0])
    set_category_last.add(categories[-1])
    for category in categories:
        set_category.add(category)
print "set_category:\n", set_category, '\n'
print "set_category_first:\n", set_category_first, '\n'
print "set_category_last:\n", set_category_last, '\n'

set_category:
set(['ACC', 'EYESHADOW', 'PADDED', 'TOWELS', 'BEACHWEAR', 'PERFUME', 'COATS', 'FRAGRANCE', 'EAUDETOILETTE', 'CARDIGANS', 'EVERYDAY', 'SPORTSWEAR', 'SLIPS', 'GLOVES', 'CONCEALER', 'SCRUB', 'NURSING', 'HANDBAGS', 'MOBILEACC', 'LONG', 'PLUS', 'DRESSING', 'MULTIPACK', 'SHAPING', 'FOUNDATION', 'LASHES', 'HANDS', 'PARTYDRESSES', 'SWIM', 'LIPS', 'MASCARA', 'TUNICS', 'SWEATSHIRT', 'PRIMER', 'LIPLINER', 'REMOVERS', 'LIMITED', 'LIPSTICK', 'CHINOS', 'STRAIGHT', 'BOOTS', 'NAILPOLISH', 'LINGERIE', 'SNEAKERS', 'JACKETS', 'BEDROOM', 'MODERN-CLASSICS', 'MAN', 'SHORT', 'BODYLOTIONS', 'HATS', 'TOPS', 'BRIEFS', 'BATH', 'SOCKS', 'SPONGES', 'HM', 'TIES', 'BATHROOM', 'PQ', 'HIGHLIGHTER', 'BLOUSES', 'FACEMASKS', 'SWEATSHIRTS', 'ROBES', 'DENIMSHORTS', 'PURSES', 'NAILCARE', 'REGULAR', 'CORSETS', 'PUSHUP', 'MATERNITY', 'APPLICATORS', 'JUMPERS', 'EYES', 'VESTS', 'BELTS', 'TURTLENECKS', 'KIMONOS', 'SCARVES', 'MODERNCLASSICS', 'BRONZER', 'JUMPSUITS', 'SWEATPANTS', 'HIPSTERS', 'SLEEVELESS', 'JEANS', '

### Find inclusive or independent categories

In [6]:
def get_supersets(category):
    is_first_iter = True
    supersets = set()
    for serial in dict_items_info:
        metricCategoryID = dict_items_info[serial]['metricCategoryID']
        # Delete the first and last strings
        categories = metricCategoryID.split('_')[1:-1]
        if category in categories:
            if is_first_iter:
                supersets = set(categories)
                is_first_iter = False
            else:
                supersets = supersets.intersection(categories)
    return supersets

In [7]:
def get_proper_supersets(category):
    is_first_iter = True
    supersets = set()
    for serial in dict_items_info:
        metricCategoryID = dict_items_info[serial]['metricCategoryID']
        # Delete the first and last strings
        categories = metricCategoryID.split('_')[1:-1]
        if category in categories:
            if is_first_iter:
                supersets = set(categories)
                is_first_iter = False
            else:
                supersets = supersets.intersection(categories)
    return supersets.difference(set({category}))

In [8]:
get_supersets('LEGGINGS')

{'LEGGINGS'}

In [9]:
get_proper_supersets('LEGGINGS')

set()

In [10]:
# print the superset categories
superset_categories = []
for category in set_category:
    supersets = get_supersets(category)
    if len(supersets) == 1:
        print "category:", category
        superset_categories.append(category)

category: ACC
category: SPORTSWEAR
category: LONG
category: TUNICS
category: LINGERIE
category: JACKETS
category: MODERN-CLASSICS
category: SHORT
category: TOPS
category: BATH
category: SOCKS
category: HM
category: BATHROOM
category: PQ
category: VESTS
category: MODERNCLASSICS
category: JEANS
category: BOTTOMS
category: SKIRTS
category: SUITS
category: HAIR
category: LEGGINGS
category: ACCESSORIES
category: NIGHTWEAR
category: SHORTS
category: DIVIDED
category: SLIM
category: BASICS
category: SWIMWEAR
category: BLAZERS
category: TSHIRT
category: SHIRTS
category: CASUAL
category: JUMPSUIT
category: TIGHTS
category: UNDERWEAR
category: BEAUTY
category: DRESSED
category: TROUSERS
category: DRESSES
category: SLEEVED
category: HOODIES
category: SHOES
category: MIDI
category: LOOSE
category: PARTYWEAR


In [11]:
print superset_categories

['ACC', 'SPORTSWEAR', 'LONG', 'TUNICS', 'LINGERIE', 'JACKETS', 'MODERN-CLASSICS', 'SHORT', 'TOPS', 'BATH', 'SOCKS', 'HM', 'BATHROOM', 'PQ', 'VESTS', 'MODERNCLASSICS', 'JEANS', 'BOTTOMS', 'SKIRTS', 'SUITS', 'HAIR', 'LEGGINGS', 'ACCESSORIES', 'NIGHTWEAR', 'SHORTS', 'DIVIDED', 'SLIM', 'BASICS', 'SWIMWEAR', 'BLAZERS', 'TSHIRT', 'SHIRTS', 'CASUAL', 'JUMPSUIT', 'TIGHTS', 'UNDERWEAR', 'BEAUTY', 'DRESSED', 'TROUSERS', 'DRESSES', 'SLEEVED', 'HOODIES', 'SHOES', 'MIDI', 'LOOSE', 'PARTYWEAR']


### Analzing categories

In [12]:
def print_has(category, max_items):
    count = 0
    for serial in dict_items_info:
        metricCategoryID = dict_items_info[serial]['metricCategoryID']
        # Delete the first and last strings
        categories = metricCategoryID.split('_')[1:-1]
        if category in categories:
            print categories
            print dict_items_info[serial]['url']
            count = count + 1
            if count == max_items:
                return

In [13]:
print_has('TUNICS', 10)

['SHIRTS', 'BLOUSES', 'TUNICS']
http://www.hm.com/us/product/67068?article=67068-C
['SHIRTS', 'BLOUSES', 'TUNICS']
http://www.hm.com/us/product/67282?article=67282-A
['HM', 'PLUS', 'TUNICS']
http://www.hm.com/us/product/64818?article=64818-A
['HM', 'PLUS', 'TUNICS']
http://www.hm.com/us/product/68210?article=68210-C
['HM', 'PLUS', 'TUNICS']
http://www.hm.com/us/product/68210?article=68210-B
['HM', 'PLUS', 'TUNICS']
http://www.hm.com/us/product/68210?article=68210-A
['HM', 'PLUS', 'TUNICS']
http://www.hm.com/us/product/60300?article=60300-A
['HM', 'PLUS', 'TUNICS']
http://www.hm.com/us/product/60318?article=60318-A
['SHIRTS', 'BLOUSES', 'TUNICS']
http://www.hm.com/us/product/63837?article=63837-A
['SHIRTS', 'BLOUSES', 'TUNICS']
http://www.hm.com/us/product/67416?article=67416-C


* 'TUNICS' is a one-piece style.

In [14]:
print_has('SHORT', 10)

['DRESSES', 'JUMPSUITS', 'SHORT']
http://www.hm.com/us/product/68527?article=68527-A
['TOPS', 'SHORT', 'SLEEVED']
http://www.hm.com/us/product/71495?article=71495-A
['SKIRTS', 'SHORT']
http://www.hm.com/us/product/58714?article=58714-A
['TSHIRT', 'SHORT', 'SLEEVED']
http://www.hm.com/us/product/71844?article=71844-A
['TSHIRT', 'SHORT', 'SLEEVED']
http://www.hm.com/us/product/71844?article=71844-B
['SKIRTS', 'SHORT']
http://www.hm.com/us/product/65377?article=65377-A
['TOPS', 'SHORT', 'SLEEVED']
http://www.hm.com/us/product/43558?article=43558-G
['TOPS', 'SHORT', 'SLEEVED']
http://www.hm.com/us/product/43558?article=43558-F
['TOPS', 'SHORT', 'SLEEVED']
http://www.hm.com/us/product/63058?article=63058-G
['DRESSES', 'JUMPSUITS', 'SHORT']
http://www.hm.com/us/product/66963?article=66963-A


In [15]:
print_has('SHORTS', 10)

['DIVIDED', 'TROUSERS', 'SHORTS']
http://www.hm.com/us/product/29578?article=29578-I
['DIVIDED', 'TROUSERS', 'SHORTS']
http://www.hm.com/us/product/68328?article=68328-D
['DIVIDED', 'TROUSERS', 'SHORTS']
http://www.hm.com/us/product/68328?article=68328-B
['DIVIDED', 'TROUSERS', 'SHORTS']
http://www.hm.com/us/product/68328?article=68328-A
['SHORTS', 'OTHERS']
http://www.hm.com/us/product/44065?article=44065-D
['SHORTS', 'OTHERS']
http://www.hm.com/us/product/65973?article=65973-D
['SHORTS', 'OTHERS']
http://www.hm.com/us/product/65973?article=65973-C
['SHORTS', 'OTHERS']
http://www.hm.com/us/product/65973?article=65973-A
['SHORTS', 'DENIMSHORTS']
http://www.hm.com/us/product/71210?article=71210-A
['EVERYDAY', 'FASHION', 'SHORTS']
http://www.hm.com/us/product/69045?article=69045-A


* 'SHORT' is a style, and 'SHORTS' is a bottom.

In [16]:
print_has('PQ',10)

['SHOES', 'PQ']
http://www.hm.com/us/product/49302?article=49302-B
['SHOES', 'PQ']
http://www.hm.com/us/product/61758?article=61758-A
['SHOES', 'PQ']
http://www.hm.com/us/product/61758?article=61758-C
['SHOES', 'PQ']
http://www.hm.com/us/product/62386?article=62386-A
['SHOES', 'PQ']
http://www.hm.com/us/product/62386?article=62386-B
['ACCESSORIES', 'PQ']
http://www.hm.com/us/product/38523?article=38523-C
['SHOES', 'PQ']
http://www.hm.com/us/product/30123?article=30123-B
['SHOES', 'PQ']
http://www.hm.com/us/product/30123?article=30123-A
['SHOES', 'PQ']
http://www.hm.com/us/product/63389?article=63389-A
['ACCESSORIES', 'PQ']
http://www.hm.com/us/product/66243?article=66243-A


* 'PQ' means 'Premium Quality'.

In [17]:
print_has('VESTS',10)

['HM', 'MAN', 'TSHIRTS', 'VESTS']
http://www.hm.com/us/product/02779?article=02779-Q
['HM', 'MAN', 'TSHIRTS', 'VESTS']
http://www.hm.com/us/product/02779?article=02779-P
['HM', 'MAN', 'TSHIRTS', 'VESTS']
http://www.hm.com/us/product/02779?article=02779-K
['HM', 'MAN', 'TSHIRTS', 'VESTS']
http://www.hm.com/us/product/02779?article=02779-L
['HM', 'MAN', 'TSHIRTS', 'VESTS']
http://www.hm.com/us/product/02779?article=02779-N
['TOPS', 'VESTS', 'SLEEVELESS']
http://www.hm.com/us/product/48064?article=48064-R
['TOPS', 'VESTS', 'SLEEVELESS']
http://www.hm.com/us/product/48064?article=48064-N
['TSHIRT', 'VESTS']
http://www.hm.com/us/product/67985?article=67985-A
['TOPS', 'VESTS', 'SLEEVELESS']
http://www.hm.com/us/product/67594?article=67594-A
['TOPS', 'VESTS', 'SLEEVELESS']
http://www.hm.com/us/product/70014?article=70014-A


* 'VESTS' is either a top or a T-shirt.

In [18]:
print_has('SUITS',10)

['HM', 'MAN', 'SUITS', 'TROUSERS']
http://www.hm.com/us/product/46504?article=46504-B
['BLAZERS', 'SUITS', 'BLAZERS']
http://www.hm.com/us/product/60546?article=60546-A
['HM', 'MAN', 'SUITS', 'TROUSERS']
http://www.hm.com/us/product/65249?article=65249-A
['HM', 'MAN', 'SUITS', 'JACKETS']
http://www.hm.com/us/product/43390?article=43390-A
['BLAZERS', 'SUITS', 'BLAZERS']
http://www.hm.com/us/product/64499?article=64499-A
['BLAZERS', 'SUITS', 'BLAZERS']
http://www.hm.com/us/product/60161?article=60161-A
['BLAZERS', 'SUITS', 'BLAZERS']
http://www.hm.com/us/product/58546?article=58546-A
['BLAZERS', 'SUITS', 'BLAZERS']
http://www.hm.com/us/product/65605?article=65605-A
['HM', 'MAN', 'SUITS', 'TROUSERS']
http://www.hm.com/us/product/48222?article=48222-D
['HM', 'MAN', 'SUITS', 'JACKETS']
http://www.hm.com/us/product/53097?article=53097-A


* 'SUITS' means a style.

In [19]:
print_has('HAIR',10)

['BEAUTY', 'HAIR']
http://www.hm.com/us/product/59036?article=59036-A
['BEAUTY', 'HAIR']
http://www.hm.com/us/product/81568?article=81568-A
['BEAUTY', 'HAIR']
http://www.hm.com/us/product/11135?article=11135-A
['BEAUTY', 'HAIR']
http://www.hm.com/us/product/81574?article=81574-A
['ACCESSORIES', 'HAIR']
http://www.hm.com/us/product/67624?article=67624-G
['ACCESSORIES', 'HAIR']
http://www.hm.com/us/product/67624?article=67624-F
['ACCESSORIES', 'HAIR']
http://www.hm.com/us/product/67624?article=67624-H
['ACCESSORIES', 'HAIR']
http://www.hm.com/us/product/53992?article=53992-H
['ACCESSORIES', 'HAIR']
http://www.hm.com/us/product/82450?article=82450-I
['ACCESSORIES', 'HAIR']
http://www.hm.com/us/product/82450?article=82450-P


* 'HAIR' is a category that is related to hair.

In [20]:
print_has('DIVIDED',10)

['DIVIDED', 'JACKETS', 'COATS', 'JACKETS']
http://www.hm.com/us/product/55258?article=55258-A
['DIVIDED', 'TROUSERS', 'SHORTS']
http://www.hm.com/us/product/29578?article=29578-I
['DIVIDED', 'TROUSERS', 'SHORTS']
http://www.hm.com/us/product/68328?article=68328-D
['DIVIDED', 'TROUSERS', 'SHORTS']
http://www.hm.com/us/product/68328?article=68328-B
['DIVIDED', 'TROUSERS', 'SHORTS']
http://www.hm.com/us/product/68328?article=68328-A
['DIVIDED', 'HOODIES', 'SWEATSHIRT']
http://www.hm.com/us/product/66983?article=66983-B
['DIVIDED', 'HOODIES', 'SWEATSHIRT']
http://www.hm.com/us/product/59885?article=59885-A
['DIVIDED', 'BASICS', 'DRESSES', 'SKIRTS']
http://www.hm.com/us/product/15975?article=15975-B
['DIVIDED', 'BASICS', 'DRESSES', 'SKIRTS']
http://www.hm.com/us/product/62061?article=62061-B
['DIVIDED', 'HOODIES', 'SWEATSHIRT']
http://www.hm.com/us/product/69412?article=69412-B


* 'DIVIDED' is not a category only included in tops or bottoms.

In [21]:
print_has('TIGHTS',10)

['MATERNITY', 'WEAR', 'TIGHTS']
http://www.hm.com/us/product/47408?article=47408-B
['LINGERIE', 'SOCKS', 'TIGHTS', 'SOCKS']
http://www.hm.com/us/product/63294?article=63294-A
['LINGERIE', 'SOCKS', 'TIGHTS', 'SOCKS']
http://www.hm.com/us/product/66888?article=66888-A
['LINGERIE', 'SOCKS', 'TIGHTS', 'TIGHTS']
http://www.hm.com/us/product/63764?article=63764-A
['MATERNITY', 'WEAR', 'TIGHTS']
http://www.hm.com/us/product/40045?article=40045-A
['LINGERIE', 'SOCKS', 'TIGHTS', 'TIGHTS']
http://www.hm.com/us/product/30127?article=30127-A
['LINGERIE', 'SOCKS', 'TIGHTS', 'TIGHTS']
http://www.hm.com/us/product/16019?article=16019-A
['LINGERIE', 'SOCKS', 'TIGHTS', 'SOCKS']
http://www.hm.com/us/product/64421?article=64421-A
['LINGERIE', 'SOCKS', 'TIGHTS', 'SOCKS']
http://www.hm.com/us/product/64421?article=64421-B
['MATERNITY', 'WEAR', 'TIGHTS']
http://www.hm.com/us/product/41415?article=41415-A


* 'TIGHTS' means a tightly slim style. ('tights' = '타이즈')

In [22]:
print_has('DRESSED',10)

['TROUSERS', 'DRESSED']
http://www.hm.com/us/product/49299?article=49299-A
['SHIRTS', 'DRESSED']
http://www.hm.com/us/product/69802?article=69802-C
['SHIRTS', 'DRESSED']
http://www.hm.com/us/product/69802?article=69802-B
['SHIRTS', 'DRESSED']
http://www.hm.com/us/product/48533?article=48533-C
['SHIRTS', 'DRESSED']
http://www.hm.com/us/product/48533?article=48533-B
['SHIRTS', 'DRESSED']
http://www.hm.com/us/product/48533?article=48533-A
['TROUSERS', 'DRESSED']
http://www.hm.com/us/product/67329?article=67329-A
['TROUSERS', 'DRESSED']
http://www.hm.com/us/product/62697?article=62697-B
['TROUSERS', 'DRESSED']
http://www.hm.com/us/product/39920?article=39920-A
['TROUSERS', 'DRESSED']
http://www.hm.com/us/product/66008?article=66008-A


* 'DRESSED' means a formal style.

In [23]:
print_has('DRESSES',20)

['EVERYDAY', 'FASHION', 'DRESSES', 'JUMPSUITS']
http://www.hm.com/us/product/67942?article=67942-A
['DRESSES', 'JUMPSUITS', 'SHORT']
http://www.hm.com/us/product/68527?article=68527-A
['DRESSES', 'JUMPSUITS', 'MIDI']
http://www.hm.com/us/product/60680?article=60680-B
['DRESSES', 'JUMPSUITS', 'MIDI']
http://www.hm.com/us/product/67133?article=67133-A
['DRESSES', 'JUMPSUITS', 'SHORT']
http://www.hm.com/us/product/66963?article=66963-A
['MODERNCLASSICS', 'DRESSES']
http://www.hm.com/us/product/71537?article=71537-A
['HM', 'PLUS', 'DRESSES']
http://www.hm.com/us/product/82227?article=82227-C
['MATERNITY', 'WEAR', 'DRESSES']
http://www.hm.com/us/product/68777?article=68777-B
['MATERNITY', 'WEAR', 'DRESSES']
http://www.hm.com/us/product/68777?article=68777-A
['DRESSES', 'JUMPSUITS', 'MIDI']
http://www.hm.com/us/product/67073?article=67073-A
['DRESSES', 'JUMPSUITS', 'MIDI']
http://www.hm.com/us/product/67073?article=67073-B
['DIVIDED', 'BASICS', 'DRESSES', 'SKIRTS']
http://www.hm.com/us/produ

In [24]:
print_has('SKIRTS',20)

['SKIRTS', 'SHORT']
http://www.hm.com/us/product/58714?article=58714-A
['SKIRTS', 'SHORT']
http://www.hm.com/us/product/65377?article=65377-A
['SKIRTS', 'SHORT']
http://www.hm.com/us/product/62051?article=62051-A
['SKIRTS', 'SHORT']
http://www.hm.com/us/product/34850?article=34850-A
['SKIRTS', 'LONG']
http://www.hm.com/us/product/66731?article=66731-A
['DIVIDED', 'BASICS', 'DRESSES', 'SKIRTS']
http://www.hm.com/us/product/15975?article=15975-B
['DIVIDED', 'BASICS', 'DRESSES', 'SKIRTS']
http://www.hm.com/us/product/62061?article=62061-B
['SKIRTS', 'SHORT']
http://www.hm.com/us/product/66352?article=66352-A
['SKIRTS', 'SHORT']
http://www.hm.com/us/product/70152?article=70152-A
['SKIRTS', 'SHORT']
http://www.hm.com/us/product/68739?article=68739-B
['SKIRTS', 'SHORT']
http://www.hm.com/us/product/32578?article=32578-A
['DIVIDED', 'BASICS', 'DRESSES', 'SKIRTS']
http://www.hm.com/us/product/27458?article=27458-M
['DIVIDED', 'BASICS', 'DRESSES', 'SKIRTS']
http://www.hm.com/us/product/27458?ar

* 'SKIRTS' and 'DRESSES' are sometimes overlapped.
* If they are overlapped, we can categorize the item by its name.
* At the end of the name of an item, __'skirt'__ or __'dress'__ is used instead of 'skirts' or 'dresses'.

In [25]:
print_has('MIDI',10)

['DRESSES', 'JUMPSUITS', 'MIDI']
http://www.hm.com/us/product/60680?article=60680-B
['DRESSES', 'JUMPSUITS', 'MIDI']
http://www.hm.com/us/product/67133?article=67133-A
['DRESSES', 'JUMPSUITS', 'MIDI']
http://www.hm.com/us/product/67073?article=67073-A
['DRESSES', 'JUMPSUITS', 'MIDI']
http://www.hm.com/us/product/67073?article=67073-B
['DRESSES', 'JUMPSUITS', 'MIDI']
http://www.hm.com/us/product/62412?article=62412-A
['SKIRTS', 'MIDI']
http://www.hm.com/us/product/69991?article=69991-A
['SKIRTS', 'MIDI']
http://www.hm.com/us/product/55496?article=55496-B
['SKIRTS', 'MIDI']
http://www.hm.com/us/product/67989?article=67989-A
['SKIRTS', 'MIDI']
http://www.hm.com/us/product/59592?article=59592-C
['SKIRTS', 'MIDI']
http://www.hm.com/us/product/68303?article=68303-A


* 'MIDI' is something related to skirts. It presumably means a style.

These categories are inclusive categories.

The following categories are categories that can be classified one of the three: tops, bottoms, and one-pieces(top-bottom together).

['ACC', 'SPORTSWEAR', 'LONG', __'TUNICS'__, 'LINGERIE', __'JACKETS'__, 'MODERN-CLASSICS', 'SHORT', __'TOPS'__, 'BATH', 'SOCKS', 'HM', 'BATHROOM', 'PQ', __'VESTS'__, 'MODERNCLASSICS', __'JEANS'__, __'BOTTOMS'__, __'SKIRTS'__, 'SUITS', 'HAIR', __'LEGGINGS'__, 'ACCESSORIES', 'NIGHTWEAR', __'SHORTS'__, 'DIVIDED', 'SLIM', 'BASICS', 'SWIMWEAR', __'BLAZERS'__, __'TSHIRT'__, __'SHIRTS'__, 'CASUAL', __'JUMPSUIT'__, 'TIGHTS', 'UNDERWEAR', 'BEAUTY', 'DRESSED', __'TROUSERS'__, __'DRESSES'__, 'SLEEVED', __'HOODIES'__, 'SHOES', 'MIDI', 'LOOSE', 'PARTYWEAR']

#### Tops
* ['JACKETS', 'TOPS', 'VESTS', 'BLAZERS', 'TSHIRT', 'SHIRTS', 'HOODIES']

#### Bottoms
* ['JEANS', 'BOTTOMS', 'SKIRTS', 'LEGGINGS', 'SHORTS', 'TROUSERS']

#### Onepieces
* ['TUNICS', 'JUMPSUIT', 'DRESSES']

In [26]:
dict_tbo_categories = dict()

In [27]:
dict_tbo_categories['top'] = ['JACKETS', 'TOPS', 'VESTS', 'BLAZERS', 'TSHIRT', 'SHIRTS', 'HOODIES']

In [28]:
dict_tbo_categories['bottom'] = ['JEANS', 'BOTTOMS', 'SKIRTS', 'LEGGINGS', 'SHORTS', 'TROUSERS']

In [29]:
dict_tbo_categories['onepiece'] = ['TUNICS', 'JUMPSUIT', 'DRESSES']

In [30]:
print dict_tbo_categories

{'top': ['JACKETS', 'TOPS', 'VESTS', 'BLAZERS', 'TSHIRT', 'SHIRTS', 'HOODIES'], 'onepiece': ['TUNICS', 'JUMPSUIT', 'DRESSES'], 'bottom': ['JEANS', 'BOTTOMS', 'SKIRTS', 'LEGGINGS', 'SHORTS', 'TROUSERS']}


### Serialization

In [63]:
import pickle
f = open('pickle_data/dict_tbo_categories.p', 'wb')
pickle.dump(dict_tbo_categories, f)
f.close()

In [31]:
import pickle
f = open('pickle_data/dict_tbo_categories.p', 'rb')
dict_tbo_categories = pickle.load(f)
f.close()

# 3. Are tops, bottoms, and onepieces mutually exclusive?

In [69]:
from interface import *

In [70]:
dict_tbo_categories = import_tbo_categories()

In [71]:
dict_tbo_categories['top']

['JACKETS', 'TOPS', 'VESTS', 'BLAZERS', 'TSHIRT', 'SHIRTS', 'HOODIES']

In [72]:
dict_tbo_categories['bottom']

['JEANS', 'BOTTOMS', 'SKIRTS', 'LEGGINGS', 'SHORTS', 'TROUSERS']

In [73]:
dict_tbo_categories['onepiece']

['TUNICS', 'JUMPSUIT', 'DRESSES']

In [74]:
pf = open('pickle_data/dict_all_items_info.p', 'rb')
dict_items_info = pickle.load(pf)
pf.close()

In [75]:
def get_categories(item_info):
    metricCategoryID = item_info['metricCategoryID']
    # Delete the first and last strings
    categories = metricCategoryID.split('_')[1:-1]
    return categories

In [76]:
get_categories(dict_items_info[dict_items_info.keys()[0]])

['HM', 'MAN', 'TSHIRTS', 'VESTS']

In [57]:
def print_categories_with_tbo_num(tbo_num_to_print):
    count = 0
    for serial in dict_items_info:
        dict_items_info[serial]['tbo'] = set()
        categories = get_categories(dict_items_info[serial])
        for category in categories:
            if category in dict_tbo_categories['top']:
                dict_items_info[serial]['tbo'].add('top')
            if category in dict_tbo_categories['bottom']:
                dict_items_info[serial]['tbo'].add('bottom')
            if category in dict_tbo_categories['onepiece']:
                dict_items_info[serial]['tbo'].add('onepiece')
        if len(dict_items_info[serial]['tbo']) == tbo_num_to_print:
            print "TBO       :", dict_items_info[serial]['tbo']
            print "categories:", categories
            print "name:", dict_items_info[serial]['name']
            print "URL       :", dict_items_info[serial]['url']
            print '\n'
            count = count + 1
    print "Count:", count 

In [52]:
print_categories_with_tbo_num(2)

TBO       : set(['top', 'onepiece'])
categories: ['SHIRTS', 'BLOUSES', 'TUNICS']
name: Embroidered Cotton Tunic
URL       : http://www.hm.com/us/product/67068?article=67068-C


TBO       : set(['onepiece', 'bottom'])
categories: ['DIVIDED', 'BASICS', 'DRESSES', 'SKIRTS']
name: Jersey Skirt
URL       : http://www.hm.com/us/product/15975?article=15975-B


TBO       : set(['onepiece', 'bottom'])
categories: ['DIVIDED', 'BASICS', 'DRESSES', 'SKIRTS']
name: Jersey Dress
URL       : http://www.hm.com/us/product/62061?article=62061-B


TBO       : set(['onepiece', 'bottom'])
categories: ['DIVIDED', 'BASICS', 'DRESSES', 'SKIRTS']
name: Maxi Dress
URL       : http://www.hm.com/us/product/27458?article=27458-M


TBO       : set(['onepiece', 'bottom'])
categories: ['DIVIDED', 'BASICS', 'DRESSES', 'SKIRTS']
name: Maxi Dress
URL       : http://www.hm.com/us/product/27458?article=27458-N


TBO       : set(['onepiece', 'bottom'])
categories: ['BASICS', 'DRESSES', 'SKIRTS']
name: Ribbed Jersey Dress
U

* The 93 items are more than one TBO categories.

In [53]:
print_categories_with_tbo_num(3)

Count: 0


###  Only for counting


In [41]:
def count_categories_with_tbo_num(tbo_num_to_print):
    count = 0
    for serial in dict_items_info:
        dict_items_info[serial]['tbo'] = set()
        categories = get_categories(dict_items_info[serial])
        for category in categories:
            if category in dict_tbo_categories['top']:
                dict_items_info[serial]['tbo'].add('top')
            if category in dict_tbo_categories['bottom']:
                dict_items_info[serial]['tbo'].add('bottom')
            if category in dict_tbo_categories['onepiece']:
                dict_items_info[serial]['tbo'].add('onepiece')
        if len(dict_items_info[serial]['tbo']) == tbo_num_to_print:
            count = count + 1
    print "Count:", count 

In [46]:
count_categories_with_tbo_num(0)

Count: 3290


In [43]:
count_categories_with_tbo_num(1)

Count: 5081


In [44]:
count_categories_with_tbo_num(2)

Count: 93


In [45]:
count_categories_with_tbo_num(3)

Count: 0


#### Reult: The number of items that has particular numbers of TBO categories
* 0 TBO categories: 3290 (These are like accesories, shoes, hats, etc..)
* 1 TBO category: 5081
* 2 TBO categories: 93
* 3 TBO categories: 0

### Preventing from having 2 TBO categories

In [54]:
print_categories_with_tbo_num(2)

TBO       : set(['top', 'onepiece'])
categories: ['SHIRTS', 'BLOUSES', 'TUNICS']
name: Embroidered Cotton Tunic
URL       : http://www.hm.com/us/product/67068?article=67068-C


TBO       : set(['onepiece', 'bottom'])
categories: ['DIVIDED', 'BASICS', 'DRESSES', 'SKIRTS']
name: Jersey Skirt
URL       : http://www.hm.com/us/product/15975?article=15975-B


TBO       : set(['onepiece', 'bottom'])
categories: ['DIVIDED', 'BASICS', 'DRESSES', 'SKIRTS']
name: Jersey Dress
URL       : http://www.hm.com/us/product/62061?article=62061-B


TBO       : set(['onepiece', 'bottom'])
categories: ['DIVIDED', 'BASICS', 'DRESSES', 'SKIRTS']
name: Maxi Dress
URL       : http://www.hm.com/us/product/27458?article=27458-M


TBO       : set(['onepiece', 'bottom'])
categories: ['DIVIDED', 'BASICS', 'DRESSES', 'SKIRTS']
name: Maxi Dress
URL       : http://www.hm.com/us/product/27458?article=27458-N


TBO       : set(['onepiece', 'bottom'])
categories: ['BASICS', 'DRESSES', 'SKIRTS']
name: Ribbed Jersey Dress
U

In [62]:
def print_categories_with_tbo_num_for_name_last_word(tbo_num_to_print):
    set_name_last_word = set()
    
    for serial in dict_items_info:
        dict_items_info[serial]['tbo'] = set()
        categories = get_categories(dict_items_info[serial])
        for category in categories:
            if category in dict_tbo_categories['top']:
                dict_items_info[serial]['tbo'].add('top')
            if category in dict_tbo_categories['bottom']:
                dict_items_info[serial]['tbo'].add('bottom')
            if category in dict_tbo_categories['onepiece']:
                dict_items_info[serial]['tbo'].add('onepiece')
        if len(dict_items_info[serial]['tbo']) == tbo_num_to_print:
            name_last_word = dict_items_info[serial]['name'].split()[-1]
            set_name_last_word.add(name_last_word)
    print "set_name_last_word:", set_name_last_word 

In [63]:
print_categories_with_tbo_num_for_name_last_word(2)

set_name_last_word: set(['Skirt', 'Jumpsuit', 'Tunic', 'Inserts', 'Dress'])


In [67]:
def print_categories_with_tbo_num_for_name_first_word(tbo_num_to_print):
    set_name_first_word = set()
    
    for serial in dict_items_info:
        dict_items_info[serial]['tbo'] = set()
        categories = get_categories(dict_items_info[serial])
        for category in categories:
            if category in dict_tbo_categories['top']:
                dict_items_info[serial]['tbo'].add('top')
            if category in dict_tbo_categories['bottom']:
                dict_items_info[serial]['tbo'].add('bottom')
            if category in dict_tbo_categories['onepiece']:
                dict_items_info[serial]['tbo'].add('onepiece')
        if len(dict_items_info[serial]['tbo']) == tbo_num_to_print:
            name_first_word = dict_items_info[serial]['name'].split()[0]
            set_name_first_word.add(name_first_word)
    print "set_name_first_word:", set_name_first_word 

In [68]:
print_categories_with_tbo_num_for_name_first_word(2)

set_name_first_word: set(['Hooded', 'Long', 'Maxi', 'Circle', 'Cr\xc3\xaape', 'Pencil', 'Knee-length', 'Fine-knit', 'T-shirt', 'Jersey', 'V-neck', 'Striped', 'Short', 'Tunic', 'Sleeveless', 'Ribbed', 'Lyocell', 'Lyocell-blend', 'Fitted', 'Long-sleeved', 'Strapless', 'Chiffon', 'Embroidered', 'Sweatshirt', 'Jumpsuit', 'Short-sleeved', 'Cotton'])


__My conclusion__: ignore the items with more than one TBO categries.

## 4. Add the TBO category

In [69]:
from interface import *

In [70]:
dict_tbo_categories = import_tbo_categories()

In [74]:
pf = open('pickle_data/dict_all_items_info.p', 'rb')
dict_items_info = pickle.load(pf)
pf.close()

In [75]:
def get_categories(item_info):
    metricCategoryID = item_info['metricCategoryID']
    # Delete the first and last strings
    categories = metricCategoryID.split('_')[1:-1]
    return categories

In [78]:
set_tbo = set()
count = 0
serials_to_delete = []
for serial in dict_items_info:
    categories = get_categories(dict_items_info[serial])
    for category in categories:
        if category in dict_tbo_categories['top']:
            set_tbo.add('top')
        if category in dict_tbo_categories['bottom']:
            set_tbo.add('bottom')
        if category in dict_tbo_categories['onepiece']:
            set_tbo.add('onepiece')
    if len(set_tbo) == 1: # Dealing with items with only one TBO category.
        dict_items_info[serial]['tbo'] = list(set_tbo)[0]
        count = count + 1
    else:
        serials_to_delete.append(serial)
    set_tbo.clear()
print "count items with 1 TBO:", count

for serial in serials_to_delete:
    del dict_items_info[serial]

print "len(dict_items_info):", len(dict_items_info)

count items with 1 TBO: 5081
len(dict_items_info): 5081


### Test

In [79]:
import random

In [146]:
random.choice(dict_items_info.items())

('63358-A',
 {'color': 'Black',
  'image': {'item': 'http://lp.hm.com/hmprod?set=key[source],value[/model/2016/E00 0453000 001 48 1528.jpg]&set=key[rotate],value[]&set=key[width],value[]&set=key[height],value[]&set=key[x],value[]&set=key[y],value[]&set=key[type],value[STILL_LIFE_FRONT]&set=key[hmver],value[1]&call=url[file:/product/large]',
   'model': 'http://lp.hm.com/hmprod?set=key[source],value[/environment/2017/8HZ_0452_013R.jpg]&set=key[rotate],value[0]&set=key[width],value[4015]&set=key[height],value[4695]&set=key[x],value[0]&set=key[y],value[-1]&set=key[type],value[FASHION_FRONT]&set=key[hmver],value[0]&call=url[file:/product/large]'},
  'metricCategoryID': 'MEN_CASUAL_JACKETS_COATS_M071',
  'name': 'Jacket with Stand-up Collar',
  'tbo': 'top',
  'url': 'http://www.hm.com/us/product/63358?article=63358-A',
  'who': 'Men'})

### Serialization

# 5. Count how many items to extract

In [69]:
from interface import *

In [70]:
dict_tbo_categories = import_tbo_categories()

In [147]:
def count_has_in(categories, dict_items_info, who):
    # who: 'Women', or 'Men'
    count = 0
    for serial in dict_items_info:
        if who != dict_items_info[serial]['who']:
            continue
        metricCategoryID = dict_items_info[serial]['metricCategoryID']
        # Delete the first and last strings
        metricCategories = metricCategoryID.split('_')[1:-1]
        for category in categories:
            if category in metricCategories:
                count = count + 1
                break
    return count

### Women

In [163]:
tbo_categories = dict_tbo_categories['top'] + dict_tbo_categories['bottom'] + dict_tbo_categories['onepiece']
women = count_has_in(tbo_categories, dict_items_info, 'Women') 
print women

3690


### Men

In [165]:
men = count_has_in(tbo_categories, dict_items_info, 'Men')
print men

1391


### Women + Men

In [166]:
print men + women

5081


### Women Top

In [167]:
women_top = count_has_in(dict_tbo_categories['top'], dict_items_info, 'Women')
print women_top

1873


### Men Top

In [168]:
men_top = count_has_in(dict_tbo_categories['top'], dict_items_info, 'Men')
print men_top

922


### Women Top + Men Top

In [169]:
print women_top + men_top

2795


### Women Bottom

In [171]:
women_bottom = count_has_in(dict_tbo_categories['bottom'], dict_items_info, 'Women')
print women_bottom

994


### Men Bottom

In [173]:
men_bottom = count_has_in(dict_tbo_categories['bottom'], dict_items_info, 'Men')
print men_bottom

469


### Women Bottom + Men Bottom

In [174]:
print women_bottom + men_bottom

1463


### Women Onepiece

In [176]:
women_onepiece = count_has_in(dict_tbo_categories['onepiece'], dict_items_info, 'Women')
print women_onepiece

823


### Men Onepiece

In [177]:
men_onepiece = count_has_in(dict_tbo_categories['onepiece'], dict_items_info, 'Men')
print men_onepiece

0


### Women Onepiece + Men Onepiece

In [178]:
print women_onepiece + men_onepiece

823


### Check

In [179]:
women == (women_top + women_bottom + women_onepiece)

True

In [180]:
men == (men_top + men_bottom + men_onepiece)

True

### Result
* The number of items in the TBO categories: 5081 (Women: 3690, Men: 1391) (5-11\_category\_analysis\_for\_all\_items.ipynb)
  * The number of tops: 2795 (Women: 1873, Men: 922)
  * The number of bottoms: 1463 (Women: 994, Men: 469)
  * The number of onepieces: 823 (Women: 823, Men: 0)